Add cronjob to compress and remove files each day

abottchen opened the pull request:
Prior to this commit, all json files were stored in text format, which consumed a great deal of disk space.  This adds a new cronjob per metrics_type that will tar and bzip2 the files each day, then remove them.  The dramatic disk space savings allows the retention rate to be extended to 90 days.

abottchen committed:
Add cronjob to compress and remove files each day

npwalker commented on manifests/pe_metric.pp:
```diff
@@ -46,6 +46,13 @@
+    ensure  => $metric_ensure,
+    user    => 'root',
+    hour    => '1',
+    command => "export DIR='${output_dir}' SVC=${metrics_type} ; find \"\$DIR/\$SVC\" -type f \! -name \"*.bz2\" > \"\$DIR/\$SVC.tmp\" ; xargs -a \"\$DIR/\$SVC.tmp\" tar -jcf \"\$DIR/\$SVC/\$SVC-$(date +%Y%m%d).tar.bz2\" 2>/dev/null ; xargs -a \"\$DIR/\$SVC.tmp\" rm ; rm \"\$DIR/\$SVC.tmp\"",
```
wow that's more complicated than I was expecting.  I think people would like it better if we combined this with the _metrics_tidy cron though.

npwalker committed:
Refactor compression and deletion into one script

Prior to this commit, deletion of old files and compression of files
were seperate cron jobs.

After this commit, we have a single script to do both and stick with
one cron job for tidying things.

npwalker commented:
@abottchen I refactored into a script to be called from a single cron job.  However, running the script more than once per day results in overriding the previous output which seems undesirable

npwalker commented:
@abottchen @jarretlavallee @reidmv Can we seamlessly unzip these files in the visualizer?  I don't want to merge this until we can easily use the visualizer with data produced from this.

abottchen commented:
@npwalker It looks to me as though it would be trivial to update the visualizer.  If nothing else, we could just add a decompress just before the files are read in at https://github.com/puppetlabs/puppetserver-metrics-viz/blob/master/view-in-grafana.sh#L37.

Commit:
Change paths in tidy compression

Prior to this commit the tar in tidy would embed the absolute path of
the json files. Since paths can change, it is worth doing a relative
path when compressing the files.

This commit changes to relative paths for the tidy script

Signed-off-by: Nick Walker <nick.walker@puppetlabs.com>

jarretlavallee commented:
Since I cannot modify this PR, I have included a patch below. This simple patch changes the tarball from using absolute paths to relative paths. This reduces effort when extracting the tars. 

~~~
From 02cb1e2fb78367b0deb9963d3ac2a1e40419bf0c Mon Sep 17 00:00:00 2001
From: Jarret Lavallee <jarret.lavallee@puppet.com>
Date: Thu, 30 Mar 2017 14:45:20 -0700
Subject: [PATCH] Change paths in tidy compression

Prior to this commit the tar in tidy would embed the absolute path of
the json files. Since paths can change, it is worth doing a relative
path when compressing the files.

This commit changes to relative paths for the tidy script
---
 templates/tidy_cron.epp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/templates/tidy_cron.epp b/templates/tidy_cron.epp
index ea12343..450cf91 100644
--- a/templates/tidy_cron.epp
+++ b/templates/tidy_cron.epp
@@ -11,7 +11,8 @@ RETENTION_DAYS=<%= $retention_days %>;
 find "$DIR" -type f -mtime $RETENTION_DAYS -delete

 #compress files
-find "$DIR" -type f ! -name "*.bz2" > "$DIR.tmp";
+cd "$DIR"
+find . -type f ! -name "*.bz2" > "$DIR.tmp";
 xargs -a "$DIR.tmp" tar -jcf "$DIR/$METRICS_TYPE-$(date +%Y%m%d).tar.bz2" 2>/dev/null;
 xargs -a "$DIR.tmp" rm ;
 rm "$DIR.tmp";
--
~~~

npwalker commented:
@jarretlavallee let me know if you can push that change up to this PR and I'll merge it.  Nevermind i figured out how to apply the patch.

npwalker committed:
Include hour, minute, and second in compressed tarball name

Prior to this commit, running the script to compress metrics files
twice in one day would result in overwriting the existing tarball.
If you want to run the script manually for debugging or any other
reason, losing data would be a not preferred side effect.

After this commit, the tarball name includes the hour, minute and
second to avoid overwriting the tarball if run twice in one day.

jarretlavallee commented:
@npwalker I see you added the initial commit. I added another that changed the cronjob to only run once daily as it was overwriting my data in the lab.

Commit:
Only run the tidy cron jobs once daily

    Prior to this commit the tidy cron jobs would run every minute of 2am,
    which overwrites the tarball initially created by the script every
    minute. This commit starts the cron job only at 2:00am instead.

npwalker commented on manifests/pe_metric.pp:
```diff
@@ -55,6 +55,7 @@
     ensure  => $metric_ensure,
     user    => 'root',
     hour    => '2',
+    minute  => '0',
```
I think I've made this mistake with pretty much every cron job I've ever made with Puppet.
